Code/Engine/InputManager.h

```diff
@@ -5,100 +5,140 @@
 
 namespace sad
 {
-
     class InputManager {
 
     public:
         static sad::InputManager& GetInstance();
         InputManager(const InputManager&) = delete;
-        float ControllerDeadZone;
+
+        int UpdateCounter = 0;
+        void IncrementUpdateCounter()
+        {
+            UpdateCounter++;
+        }
 
         // Keyboard Events
-        bool GetKey(SDL_Scancode key); // Returns true if key is held
-        bool GetKeyPressed(SDL_Scancode key); // Returns true if key is pressed
-        bool GetKeyReleased(SDL_Scancode key); // Returns true if key is released
+        std::unordered_map<SDL_Scancode, bool> keyboardState;
+        std::unordered_map<SDL_Scancode, int>  keyboardUpdateFrames;
 
-        bool GetMouseButton(SDL_MouseButtonEvent button); // Returns true if mouse button is held
-        bool GetMouseButtonPressed(SDL_MouseButtonEvent button); // Returns true if mouse button is pressed
-        bool GetMouseButtonReleased(SDL_MouseButtonEvent button); // Returns true if mouse button is released
+        /**
+         * @brief Catches keyboard events and updates their states
+         * @param event 
+        */
+        void CatchKeyboardEvents(SDL_Event& event);
 
-        const Uint8* CurrentKeyboardStates = SDL_GetKeyboardState(nullptr);
-        std::map<SDL_Scancode, bool> KeyboardStates;
+        /**
+         * @brief Returns true if a keyboard key is held.
+         * @param key scancode
+         * @return 
+        */
+        bool GetKey(SDL_Scancode key);
 
         /**
-         * @brief Saves the down state of a key.
-         * @param key
-         * @param pressed
+         * @brief Returns true if a keyboard key is pressed during the current frame
+         * @param key scancode
+         * @return
         */
-        void UpdateKeyboardState(SDL_Scancode key, bool pressed)
-        {
-            std::map<SDL_Scancode, bool>::iterator iter = KeyboardStates.find(key);
-            if (iter != KeyboardStates.end())
-            {
-                iter->second = pressed;
-                return;
-            }
-            KeyboardStates.insert({ key, pressed });
-        }
+        bool GetKeyPressed(SDL_Scancode key);
 
         /**
-         * @brief Returns true if a key is down in the previous frame.
-         * @param key
+         * @brief Returns true if a keyboard key is released during the current frame
+         * @param key scancode
          * @return
         */
-        bool GetKeyboardState(SDL_Scancode key)
-        {
-            std::map<SDL_Scancode, bool>::iterator iter = KeyboardStates.find(key);
-            if (iter != KeyboardStates.end())
-            {
-                return iter->second;
-            }
-            return false;
+        bool GetKeyReleased(SDL_Scancode key);
+
+
+        // Mouse Events
+        SDL_Point MousePosition;
+        bool mouseState[6];
+        int  mouseUpdateFrames[6];
```

Comment:
Are these supposed to be public? Do we want other classes to be able to modify the `MousePosition` and/or `MouseState`?

If these should be private, just make sure we refactor it to have `m_` prefixing the variable names.

```
private:
    bool m_MouseState[6];
    int m_MouseUpdateFrames[6];

// ...etc.
```